src/python/pants/rules/core/strip_source_roots.py

```diff
@@ -0,0 +1,98 @@
+
+  The field `sentinel_file` is used to determine the source root for the files to be stripped. This
+  assumes that every file shares the same source root, which should be true in practice as the
+  `sources` field for a target always has files in the same source root (outside of aggregate
+  targets, which we don't expect to ever strip the source root from). See
+  https://github.com/pantsbuild/pants/pull/9112#discussion_r377999025 for more context on this
+  design.
+  """
+  snapshot: Snapshot
+  sentinel_file: str
+
+  def determine_source_root(self, *, source_root_config: SourceRootConfig) -> str:
+    source_roots_object = source_root_config.get_source_roots()
+    source_root = source_roots_object.safe_find_by_path(self.sentinel_file)
+    if source_root is not None:
+      return cast(str, source_root.path)
+    if source_root_config.options.unmatched == "fail":
+      raise NoSourceRootError(f"Could not find a source root for `{self.sentinel_file}`.")
+    # Otherwise, create a source root by using the parent directory.
+    return PurePath(self.sentinel_file).parent.as_posix()
+
+
+@rule
+async def strip_source_roots_from_snapshot(
+  request: StripSourceRootsRequest, source_root_config: SourceRootConfig,
+) -> SourceRootStrippedSources:
+  """Removes source roots from a snapshot,
+  e.g. `src/python/pants/util/strutil.py` -> `pants/util/strutil.py`.
+  """
+  source_root = request.determine_source_root(source_root_config=source_root_config)
+  resulting_digest = await Get[Digest](
+    DirectoryWithPrefixToStrip(
+      directory_digest=request.snapshot.directory_digest,
+      prefix=source_root,
+    )
+  )
+  resulting_snapshot = await Get[Snapshot](Digest, resulting_digest)
+  return SourceRootStrippedSources(snapshot=resulting_snapshot)
+
+
+@rule
+async def strip_source_roots_from_target(
+  hydrated_target: HydratedTarget,
+) -> SourceRootStrippedSources:
+  """Remove source roots from a target (depending upon the target type), e.g.
+  `src/python/pants/util/strutil.py` -> `pants/util/strutil.py`.
+  """
+  target_adaptor = hydrated_target.adaptor
+
+  # TODO: make TargetAdaptor return a 'sources' field with an empty snapshot instead of raising to
+  # simplify the hasattr() checks here!
+  if not hasattr(target_adaptor, 'sources'):
+    return SourceRootStrippedSources(snapshot=EMPTY_SNAPSHOT)
+
+  # Loose `Files`, as opposed to `Resources` or `Target`s, have no (implied) package
+  # structure and so we do not remove their source root like we normally do, so that filesystem
+  # APIs may still access the files. See pex_build_util.py's `_create_source_dumper`.
+  if target_adaptor.type_alias == Files.alias():
+    return SourceRootStrippedSources(snapshot=target_adaptor.sources.snapshot)
+
+  # NB: The sentinel file does not need to actually exist. It is solely used to determine what the
+  # source root is for the target.
+  sentinel_file = PurePath(hydrated_target.address.spec_path, "SENTINEL").as_posix()
```

Comment:
Use of the SENTINEL suffix, pollutes error messages, for example. I think here you can just pass `hydrated_target.address.spec_path`? That's what the old code does, anyway.

Reply:
The issue is that `spec_path` is a path to the parent directory, not to the actual file. We need a path to a representative file.

We can't rely on being able to say `target_adaptor.sources.snapshot.files[0]` because the snapshot might be empty. Also, the target might use rglobs and we wouldn't be able to guarantee that `snapshot.files[0]` is in the root of the target vs. a subdirectory.

We could `await Get[BuildFileAddress](Address)`, but that has a performance hit (even if it's minor).